dev/tasks/jars/build-cpp-osx.sh

```diff
@@ -29,21 +31,37 @@ pushd cpp
           -DARROW_GANDIVA=ON \
           -DARROW_GANDIVA_JAVA=ON \
           -DARROW_GANDIVA_STATIC_LIBSTDCPP=ON \
+          -DARROW_ORC=ON \
+          -DARROW_JNI=ON \
+          -DARROW_PLASMA=ON \
+          -DARROW_PLASMA_JAVA_CLIENT=ON \
           -DARROW_BUILD_TESTS=ON \
           -DARROW_BUILD_UTILITIES=OFF \
-          -DPARQUET_BUILD_ENCRYPTION=OFF \
-          -DARROW_PARQUET=OFF \
-          -DARROW_FILESYSTEM=OFF \
-          -DARROW_DATASET=OFF \
+          -DPARQUET_REQUIRE_ENCRYPTION=OFF \
+          -DARROW_PARQUET=ON \
+          -DPARQUET_BUILD_EXAMPLES=OFF \
+          -DPARQUET_BUILD_EXECUTABLES=OFF \
+          -DARROW_FILESYSTEM=ON \
+          -DARROW_DATASET=ON \
           -DARROW_BOOST_USE_SHARED=OFF \
           -DARROW_PROTOBUF_USE_SHARED=OFF \
           -DARROW_GFLAGS_USE_SHARED=OFF \
-          -DARROW_OPENSSL_USE_SHARED=OFF"
+          -DARROW_OPENSSL_USE_SHARED=OFF \
+          -DARROW_BROTLI_USE_SHARED=OFF \
+          -DARROW_BZ2_USE_SHARED=OFF \
+          -DARROW_GRPC_USE_SHARED=OFF \
+          -DARROW_LZ4_USE_SHARED=OFF \
+          -DARROW_SNAPPY_USE_SHARED=OFF \
+          -DARROW_THRIFT_USE_SHARED=OFF \
+          -DARROW_UTF8PROC_USE_SHARED=OFF \
+          -DARROW_ZSTD_USE_SHARED=OFF"
 
     cmake $CMAKE_FLAGS ..
     make -j4
     ctest
 
     cp -L release/libgandiva_jni.dylib $GITHUB_WORKSPACE/arrow/dist
+    cp -L release/libarrow_dataset_jni.dylib $GITHUB_WORKSPACE/arrow/dist
```

Comment:
Please factor out the `GITHUB_WORKSPACE` env variable and use arguments instead so we can more easily call this file locally, for example using arguments like in [cpp_build.sh](https://github.com/apache/arrow/blob/master/ci/scripts/cpp_build.sh#L22-L24)